### PR TITLE
Make Crate rules compatible with crates.io

### PR DESCRIPTION
**Pull request by vmax**
## What is the goal of this PR?

Initial implementation of `assemble_crate` and `deploy_crate` done in #322 is not fully compatible with `crates.io` registry and therefore needs to be made compatible.

## What are the changes implemented in this PR?

* Make `description`, `license` and `repository` fields mandatory
* Correctly handle empty `documentation`
* Validate `homepage`, `repository` and `documentation` fields to be valid URLs
* Validate `keywords` list to be valid
* Account for possibility of validation errors on deployment

### CODE REVIEW

File: crates/CrateDeployer.kt [hunk 1]
```diff
@@ -83,7 +84,16 @@ class CrateDeployer : Callable<Unit> {
             .put(crateContent)
             .array()
 
-        httpPut(repoUrl, token, payload)
+        val response = httpPut(repoUrl, token, payload).parseAsString()
+        if (response.isNotBlank()) {
```
Comment (vmax) [inline crates/CrateDeployer.kt:88]: Correct responses are blank and cannot be parsed as JSON

File: crates/rules.bzl [hunk 2]
```diff
@@ -53,12 +60,15 @@ def _assemble_crate_impl(ctx):
         "--keywords", ";".join(ctx.attr.keywords),
         "--categories", ";".join(ctx.attr.categories),
         "--description", ctx.attr.description,
-        "--documentation", ctx.attr.documentation,
         "--homepage", ctx.attr.homepage,
         "--license", ctx.attr.license,
         "--repository", ctx.attr.repository,
         "--deps", ";".join(["{}={}".format(k, v) for k, v in ctx.attr.deps.items()]),
     ]
+    if ctx.attr.documentation != "":
```
Comment (vmax) [inline crates/rules.bzl:68]: `""` is default value for when a Bazel attribute with type `string` is not assigned

File: crates/rules.bzl [hunk 3]
```diff
@@ -119,9 +134,14 @@ assemble_crate = rule(
             doc = """Project categories""",
         ),
         "license": attr.string(
-            doc = """License of the project""",
+            mandatory = True,
+            doc = """
```
Comment (vmax) [inline crates/rules.bzl:138]: Doc blurbs are copied from Cargo documentation and relevant field is linked

File: crates/CrateAssembler.kt [hunk 4]
```diff
@@ -178,14 +180,16 @@ class CrateAssembler : Callable<Unit> {
                 obj.set("version_req", depVer)
                 obj.set("target", Json.NULL)
                 obj.set("kind", "normal")
-                obj.set("registry", "https://github.com/rust-lang/crates.io-index")
+                obj.set("registry", "")
```
Comment (vmax) [inline crates/CrateAssembler.kt:183]: `crates.io` discards non-empty values here

**Review by lolski — APPROVED** [review]

**Pull request merged**